File: Transformers/MaybeSecureValueDataTransformer.m
//
//  MaybeSecureValueDataTransformer.m
//  Cog
//
//  Created by Christopher Snowhill on 9/10/25.
//
//  This stupid wrapper exists because NSSecureUnarchiveFromDataTransformer requires
//  macOS 10.14, but the necessary elements to recreate it already exist on 10.13,
//  which is still our current minimum target. Beside which, this allows us to limit
//  the exact types of objects that are allowed to be decoded from the dictionary to
//  just those that are currently used for metadata storage.

#import "MaybeSecureValueDataTransformer.h"

@implementation MaybeSecureValueDataTransformer

+ (Class)transformedValueClass {
	return [NSData class];
}
+ (BOOL)allowsReverseTransformation {
	return YES;
}

// Convert a dictionary of metadata values into data
- (id)transformedValue:(id)value {
	if(value == nil) return nil;

	if(![value isKindOfClass:[NSDictionary class]]) return nil;

	NSError *error = nil;
	NSData *ret = [NSKeyedArchiver archivedDataWithRootObject:value requiringSecureCoding:YES error:&error];
	if(error) {
		NSLog(@"Error archiving secure data: %@", error);
		return nil;
	}
	return ret;
}

// Unarchive a dictionary of metadata values from data
- (id)reverseTransformedValue:(id)value {
	if(value == nil) return nil;

	if(![value isKindOfClass:[NSData class]]) return nil;

	NSError *error = nil;
	NSKeyedUnarchiver *unarchiver = [[NSKeyedUnarchiver alloc] initForReadingFromData:value error:&error];
	if(error) {
		NSLog(@"Error initializing unarchiver for data: %@", error);
		return nil;
	}
	unarchiver.requiresSecureCoding = YES;
	NSSet *classes = [NSSet setWithArray:@[[NSDictionary class], [NSArray class], [NSString class], [NSNumber class], [NSDate class]]];
	id ret = [unarchiver decodeObjectOfClasses:classes forKey:@"root"];
	if(!ret) {
		NSLog(@"Error decoding dictionary from data");
	}
	return ret;
}
@end

File: Transformers/MaybeSecureValueDataTransformer.swift
import Foundation

@objc(MaybeSecureValueDataTransformer)
class MaybeSecureValueDataTransformer: ValueTransformer {
    override class func transformedValueClass() -> AnyClass {
        return NSData.self
    }

    override class func allowsReverseTransformation() -> Bool {
        return true
    }

    override func transformedValue(_ value: Any?) -> Any? {
        if value == nil {
            return nil
        }

        if !(value is NSDictionary) {
            return nil
        }

        do {
            let ret = try NSKeyedArchiver.archivedData(withRootObject: value!, requiringSecureCoding: true)
            return ret
        } catch {
            NSLog("Error archiving secure data: \(error)")
            return nil
        }
    }

    override func reverseTransformedValue(_ value: Any?) -> Any? {
        if value == nil {
            return nil
        }

        if !(value is NSData) {
            return nil
        }

        if let value = value as? NSData {
            do {
                let unarchiver = try NSKeyedUnarchiver(forReadingFrom: value as Data)
                unarchiver.requiresSecureCoding = true
                let classes = [NSDictionary.self, NSArray.self, NSString.self, NSNumber.self, NSDate.self]
                let ret = unarchiver.decodeObject(of: classes, forKey: "root")
                if ret == nil {
                    NSLog("Error decoding dictionary from data")
                }
                return ret
            } catch {
                NSLog("Error initializing unarchiver for data: \(error)")
                return nil
            }
        }

        return nil
    }
}
